File: swift/Sources/AlloyAudio/DSP/AdditiveGenerator.swift
import Foundation

/// Sine partial bank (drawbar organs are a 9-partial preset of this).
/// Sustained kind: never self-finishes; the voice TVA ends the note.
/// Twin of web src/dsp/additive-generator.ts (canonical).
public struct AdditivePartial: Codable {
    public let ratio: Double
    public let level: Double

    public init(ratio: Double, level: Double) {
        self.ratio = ratio
        self.level = level
    }
}

public final class AdditiveGenerator: ToneGenerator {
    private let partials: [AdditivePartial]
    private let sampleRate: Double
    private var phases: [Double]
    private var frequency = 0.0
    private var amp = 0.0
    private var keyed = false

    public init(partials: [AdditivePartial], sampleRate: Double) {
        self.partials = partials
        self.sampleRate = sampleRate
        phases = [Double](repeating: 0, count: partials.count)
    }

    public var finished: Bool { false }

    public func noteOn(midi: Int, velocity: Double) {
        frequency = Pitch.frequency(midi: midi)
        amp = velocity
        keyed = true
        for i in phases.indices {
            phases[i] = 0
        }
    }

    public func noteOff() {
        // Intentionally empty: no intrinsic envelope to key up.
    }

    public func render(into out: inout [Float], frames: Int) {
        guard keyed else { return }
        for n in 0..<frames {
            var sample = 0.0
            for p in partials.indices {
                sample += sin(DspConstants.twoPi * phases[p]) * partials[p].level
                phases[p] += frequency * partials[p].ratio / sampleRate
                phases[p] -= phases[p].rounded(.down)
            }
            out[n] += Float(sample * amp)
        }
    }
}

File: web/packages/alloy-audio/src/dsp/additive-generator.ts
// Sine partial bank — drawbar organs are literally this (a 9-partial
// preset), and it doubles as a clean pad/bell generator. Sustained kind:
// renders until the voice's TVA (phase 1b) ends the note; noteOff is a
// no-op here. Twin: AdditiveGenerator.swift.

import { TWO_PI, type ToneGenerator } from './dsp-types.js';
import { midiToFrequency } from '../pitch.js';

export interface AdditivePartial {
  /** Frequency ratio relative to the note frequency. */
  ratio: number;
  /** Linear amplitude of this partial. */
  level: number;
}

export class AdditiveGenerator implements ToneGenerator {
  private readonly phases: number[];
  private frequency = 0;
  private pitchRatio = 1;
  private amp = 0;
  private keyed = false;

  constructor(
    private readonly partials: readonly AdditivePartial[],
    private readonly sampleRate: number,
  ) {
    this.phases = partials.map(() => 0);
  }

  /** Sustained kind: never self-finishes; the voice TVA ends the note. */
  get finished(): boolean {
    return false;
  }

  noteOn(midi: number, velocity: number): void {
    this.pitchRatio = 1;
    this.frequency = midiToFrequency(midi);
    this.amp = velocity;
    this.keyed = true;
    this.phases.fill(0);
  }

  noteOff(): void {
    // Intentionally empty: no intrinsic envelope to key up.
  }

  setPitchRatio(ratio: number): void {
    this.pitchRatio = ratio;
  }

  render(out: Float32Array, frames: number): void {
    if (!this.keyed) {
      return;
    }
    for (let n = 0; n < frames; n++) {
      let sample = 0;
      for (let p = 0; p < this.partials.length; p++) {
        sample += Math.sin(TWO_PI * this.phases[p]) * this.partials[p].level;
        this.phases[p] += (this.frequency * this.pitchRatio * this.partials[p].ratio) / this.sampleRate;
        this.phases[p] -= Math.floor(this.phases[p]);
      }
      out[n] += sample * this.amp;
    }
  }
}
